File: gil_test.java
class GilTest {
    // 空循环20亿次
    static void work() { for (long i = 0; i < 2_000_000_000L; i++); }

    public static void main(String[] args) throws Exception {
        // 串行跑两次
        long start = System.currentTimeMillis();
        work(); work();
        System.out.println("串行耗时: " + (System.currentTimeMillis() - start) + "ms");

        // 并行跑两次
        start = System.currentTimeMillis();
        Thread t1 = new Thread(GilTest::work);
        Thread t2 = new Thread(GilTest::work);
        t1.start(); t2.start(); // 同时开跑
        t1.join(); t2.join();   // 等待都跑完
        System.out.println("并行耗时: " + (System.currentTimeMillis() - start) + "ms");
    }
}

File: gil_test.cpp
#include <iostream>
#include <thread>
#include <chrono>

// 空循环20亿次，使用volatile防止编译器对循环进行优化
void work() {
    volatile long long sum = 0;
    for (long long i = 0; i < 2000000000L; ++i) {
        sum += 1;
    }
}

int main() {
    using namespace std::chrono;

    // 串行跑两次
    auto start = high_resolution_clock::now();
    work(); work();
    auto end = high_resolution_clock::now();
    std::cout << "Serial Time (C++): "
              << duration_cast<milliseconds>(end - start).count() << "ms" << std::endl;

    // 并行跑两次
    start = high_resolution_clock::now();
    std::thread t1(work);
    std::thread t2(work);
    t1.join(); t2.join();
    end = high_resolution_clock::now();
    std::cout << "Parallel Time (C++): "
              << duration_cast<milliseconds>(end - start).count() << "ms" << std::endl;

    return 0;
}



